### ISSUE TITLE
Support 4D arrays

### ISSUE THREAD

**Issue by Monnoroch** [gpujs/gpu.js#217]
The documentation says this library only supports kernels working with 1,2,3D arrays. How hard it would be to support 4D arrays?

**Comment by robertleeplummerjr** [comment]
Currently, it'd be hard.  But long term, it may be a good move.  Are you talking hypothetically, or do you have a use case where you have 4d+ arrays?

**Comment by harish2704** [comment]
* gpu.js will be very much helpful if we want to run evaluation of any pre-trained neural network models in client side.
* In the case of Neural netowk/Machine learning, most of common use case of 4D array is 2DConvolution.
  - 2DConvoulution will transform one 2D image ( or array ) into array of 2D images ( 3D array ).
* Applying 2D convolution on array of 2DImages is a example use of 4D array .

-just an info

**Comment by robertleeplummerjr** [comment]
>* gpu.js will be very much helpful if we want to run evaluation of any pre-trained neural network models in client side.

Agreed!  But we want to as well ensure that gpu.js can be used to train neural networks, and that vision will be achieved more in version 2 when OpenCL is added (node) and WebGL 2 is supported (client and node), however training does work as of now in both environments.


>* In the case of Neural netowk/Machine learning, most of common use case of 4D array is 2DConvolution.
>   * 2DConvoulution will transform one 2D image ( or 2D array ) into array of 2D images ( 3D array ).
>* Applying 2D convolution on array of 2DImages (3DArray ) will result a 4D array which is an example use of 4D array.

[Here](https://github.com/BrainJS/brain.js/blob/ff66428f828c528c1cf61379c761c1eb7f50c471/src/layer/convolution.js#L87) (brain.js) is an example of a convolution that runs on gpu.js.  While there are several layers of of looping, the result is still three dimensional: https://github.com/BrainJS/brain.js/blob/ff66428f828c528c1cf61379c761c1eb7f50c471/src/layer/convolution.js#L61

Does that help at all?

**Comment by harish2704** [comment]
Thanks. brain.js was very useful link.

Regarding Opencl: Opencl backend will be super cool feature .. I will be waiting for it..
